### PR TITLE
FEATURE: Add support for custom storage collection per asset source

### PR DESCRIPTION
**Pull request by dfeyer**
This change add a new Interface to define custom storage collection per
asset source. It contains also minor change to the current
NeosAssetProxy, this class now use the NeosAssetProxyInterface and the
AssetController use this interface to check if the current asset proxy
expose the getAsset method.

### CODE REVIEW

File: Neos.Media.Browser/Classes/Controller/AssetController.php [hunk 1]
```diff
@@ -572,7 +623,8 @@ public function updateAssetResourceAction(AssetInterface $asset, PersistentResou
             $originalFilename = $asset->getLabel();
             $this->assetService->replaceAssetResource($asset, $resource, $options);
         } catch (\Exception $exception) {
-            $this->addFlashMessage('couldNotReplaceAsset', '', Message::SEVERITY_OK, [], 1463472606);
+            $this->addFlashMessage('couldNotReplaceAsset', '', Message::SEVERITY_ERROR, [], 1463472606);
+            $this->systemLogger->logException($exception);
```
Comment (dfeyer) [inline Neos.Media.Browser/Classes/Controller/AssetController.php]: This one can be extracted to an other PR ... if required ;)

File: Neos.Media.Browser/Classes/Controller/AssetController.php [hunk 2]
```diff
@@ -294,6 +299,9 @@ public function newAction()
      */
     public function replaceAssetResourceAction(Asset $asset)
     {
+        $assetSourceIdentifier = $asset->getAssetSourceIdentifier();
+        $this->applyActiveAssetSourceToBrowserState($assetSourceIdentifier);
```
Comment (dfeyer) [inline Neos.Media.Browser/Classes/Controller/AssetController.php]: this is required to be sure, that the current asset source in the browser state, is the one attached with the current asset.

**Comment by dfeyer** [comment]
i have the bad feeling the this controller is reponsible for too much things ... fat controller :( can be a task for an other PR

**Comment by daniellienert** [comment]
@dfeyer: Would you please rebase this branch on master to give the tests the chance to run through.

**Comment by daniellienert** [comment]
@dfeyer: Would you please rebase this branch on master to give the tests the chance to run through.

**Comment by dfeyer** [comment]
@daniellienert Done and green

**Comment by robertlemke** [comment]
As far as I understand it "storage collection" in fact means Flow resource collection, right? We have to be super careful here with wording, because we have different levels which already use terms like collections, objects, resources, assets and so on with slightly different meanings.

While we introduce a new interface here, is there anything else we need to put into that interface now (which is hard to add later)? I can't think of anything yet, but maybe you do?

**Comment by jonnitto** [comment]
Ping

**Pull request closed without merging**